### PR TITLE
zh-TW: Day 1 Translations

### PR DESCRIPTION
**Pull request by henrif75**
Professional translations for missing day 1 entries #684

The diff is large mostly because of the normalization.

### CODE REVIEW

**Comment by victorhsieh** [comment]
Wow, the diff is 21K lines.  Any hints to help me narrow down to smaller ranges?

**Comment by mgeisler** [comment]
> The diff is large mostly because of the normalization.

What kind of normalization are you doing?

My guess is that you've used `mdbook-i18n-normalize`? You don't have to do this since there hasn't been any non-compatible changes made to mdbook-i18n-helpers since version 0.2.0. Infact, `mdbook-i18n-normalize` is a bit clumsy: even with https://github.com/google/mdbook-i18n-helpers/pull/93, there will be differences in how it wraps the lines compared to `msgmerge` (and also `msgcat`, which uses the same wrapping as `msgmerge`). This is part of https://github.com/google/mdbook-i18n-helpers/issues/32.

All in all, this means that you should try to submit PRs with files that have been touched lastby one of the Gettext tools: by `msgcat` or `msgmerge`.

However, I tried running the file thought `msgcat -o po/zh-TW.po po/zh-TW.po` just now, and I still get a lot of differences:

```
% git diff HEAD^ | diffstat
 zh-TW.po |13296 ++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------------
 1 file changed, 6704 insertions(+), 6592 deletions(-)
```

So the changes are not primarily because of changes to the line numbers.

Looking through `gh pr diff 1322 | bat -l patch`, I see changes to files that are outside of Day 1:

```
gh pr diff 1322 | rg '\+#:' | cut -d : -f 2 | sort -u
 src/android/aidl/changing.md
 src/android/aidl/deploy.md
...
 src/bare-metal/alloc.md
 src/bare-metal/android.md
...
 src/exercises/android/morning.md
 src/exercises/bare-metal/afternoon.md
...
 src/exercises/day-2/afternoon.md
 src/exercises/day-2/book-library.md
...
```

So I think the `zh-TV.po` file needs more fine-grained splitting.

I recently found that I can `msggrep` out all files below the `android/`, `bare-metal/`, `conconcurrency/`, and `async/` directories with this lovely incantation (in Zsh):

```sh
msggrep --location=src/{exercises/,}{async,bare-metal}{.md,"/*","/*/*","/*/*/*"} po/zh-TW.po
```

The rest (`msggrep --invert-match`) is Days 1 to 3. Getting the content of those days can be done similarly, but there are many more directories to match. Infact, now that I think of it, it might be easier to list the files explicitly to `msggrep`.

**Comment by henrif75** [comment]
> > The diff is large mostly because of the normalization.
> 
> What kind of normalization are you doing?
> 
Yes, I used the `mdbook-i18n-normalize` at the end and most of the diff comes from the new lines and line numbers. 

> However, I tried running the file thought `msgcat -o po/zh-TW.po po/zh-TW.po` just now, and I still get a lot of differences:
> 
> ```
> % git diff HEAD^ | diffstat
>  zh-TW.po |13296 ++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------------
>  1 file changed, 6704 insertions(+), 6592 deletions(-)
> ```
> 
> So the changes are not primarily because of changes to the line numbers.
> 
When merging the translations, I did against an updated messages.pot. The diffs are coming mostly because of organic changes on the English text.
I'll resubmit the PRs anyway because of the `mdbook-i18n-normalize`, but to address the big diff we could:
- Refresh the PO file first or
- Just merge the translations against the current PO file without refreshing it.
Any suggestions are welcome.

**Comment by mgeisler** [comment]
> * Just merge the translations against the current PO file without refreshing it.

That is probably the right approach: separating the use of `msgmerge` (and `mdbook-xgettext`) from the translations.

With that, the steps become

1. Regenerate `messages.pot` once. This sets POT-Creation-Date to _now_.
2. Run `msgmerge` to merge `messages.pot` into the existing translation.
3. Split the PO file into smaller chunks with `msggrep`.
4. Send the chunks to the in-house translators.
5. When translations come back, use `msgcat` to join them with the PO file. Be careful that the translations don't just end up in the back of the PO file: the `msgid` has to exist in the PO file for `msgcat` to add the `msgstr` at the right spot. If the `msgid` changed, then perhaps `msgmerge` is the right tool, I'm not 100% sure here.
6. Open a PR with the changes.

This should ensure that the POT-Creation-Date header is set correct (to the date when the translation was submitted to the in-house translators, not to today's day or to any time in-between). It should also ensure that PRs only contain small amounts of changes.

If I understand correctly, the main change here is to only extract English strings rarely: once before sending the PO files to the translators.

**Comment by henrif75** [comment]
> This should ensure that the POT-Creation-Date header is set correct (to the date when the translation was submitted to the in-house translators, not to today's day or to any time in-between). It should also ensure that PRs only contain small amounts of changes.

I had done that when I sent the file to the translators, with the difference that I sent a single file instead of the chunks. I think I can adapt your procedure by:
1. Update manually the POT-Creation-Date to the date when I refreshed the PO file with messages.pot.
2. Split the translated file into those chunks and merge them with the current zh-TW.po file, without refreshing again with messages.pot

Does it sound correct?

**Comment by henrif75** [comment]
I think now it's much more manageable?

**Comment by mgeisler** [comment]
> I think now it's much more manageable?

Yes, it looks much better now! The diff actually loads in GitHub and from a quick skim, it looks like it's only adding translations! :tada:

File: po/zh-TW.po [hunk 1]
```diff
@@ -333,99 +334,99 @@ msgstr "區塊"
...
+#: src/SUMMARY.md:121 src/modules.md:1
 msgid "Modules"
-msgstr ""
+msgstr "模組"
 
-#: src/SUMMARY.md:116 src/modules/visibility.md:1
+#: src/SUMMARY.md:122 src/modules/visibility.md:1
 msgid "Visibility"
-msgstr ""
+msgstr "瀏覽權限"
```
Comment (victorhsieh) [inline po/zh-TW.po:407]:
```suggestion
msgstr "能見度"
```

File: po/zh-TW.po [hunk 2]
```diff
@@ -463,54 +464,54 @@ msgstr "預設方法"
...
+#: src/SUMMARY.md:149
 msgid "Read and Write"
-msgstr ""
+msgstr "Read 和 Write"
 
-#: src/SUMMARY.md:144
+#: src/SUMMARY.md:150
 msgid "Drop"
-msgstr ""
+msgstr "拖放"
```
Comment (victorhsieh) [inline po/zh-TW.po:493]:
```suggestion
msgstr "Drop"
```

File: po/zh-TW.po [hunk 3]
```diff
@@ -849,17 +850,17 @@ msgstr "無界限的通道"
...
+#: src/SUMMARY.md:274
 msgid "Send"
-msgstr ""
+msgstr "傳送"
 
-#: src/SUMMARY.md:265
+#: src/SUMMARY.md:274
 msgid "Sync"
-msgstr ""
+msgstr "同步處理"
```
Comment (victorhsieh) [inline po/zh-TW.po:863]:
```suggestion
msgstr "Sync"
```

File: po/zh-TW.po [hunk 4]
```diff
@@ -869,145 +870,145 @@ msgstr "範例"
 msgid "Shared State"
 msgstr "共享狀態"
 
-#: src/SUMMARY.md:270
+#: src/SUMMARY.md:279
 msgid "Arc"
-msgstr ""
+msgstr "弧形"
```
Comment (victorhsieh) [inline po/zh-TW.po:875]:
```suggestion
msgstr "Arc"
```

File: po/zh-TW.po [hunk 5]
```diff
@@ -2013,127 +2022,141 @@ msgid ""
...
 
 #: src/welcome-day-1/what-is-rust.md:12
 msgid "firmware and boot loaders,"
-msgstr ""
+msgstr "韌體和啟動載入器"
 
 #: src/welcome-day-1/what-is-rust.md:13
 msgid "smart displays,"
-msgstr ""
+msgstr "智慧螢幕"
```
Comment (victorhsieh) [inline po/zh-TW.po:2117]:
```suggestion
msgstr "智慧螢幕、"
```

File: po/zh-TW.po [hunk 6]
```diff
@@ -2013,127 +2022,141 @@ msgid ""
...
 
 #: src/welcome-day-1/what-is-rust.md:14
 msgid "mobile phones,"
-msgstr ""
+msgstr "手機"
```
Comment (victorhsieh) [inline po/zh-TW.po:2121]:
```suggestion
msgstr "手機、"
```

File: po/zh-TW.po [hunk 7]
```diff
@@ -2013,127 +2022,141 @@ msgid ""
...
 
 #: src/welcome-day-1/what-is-rust.md:15
 msgid "desktops,"
-msgstr ""
+msgstr "電腦"
```
Comment (victorhsieh) [inline po/zh-TW.po:2125]:
```suggestion
msgstr "電腦、"
```

File: po/zh-TW.po [hunk 8]
```diff
@@ -2013,127 +2022,141 @@ msgid ""
...
 
 #: src/welcome-day-1/what-is-rust.md:16
 msgid "servers."
-msgstr ""
+msgstr "伺服器"
```
Comment (victorhsieh) [inline po/zh-TW.po:2129]:
```suggestion
msgstr "伺服器。"
```

File: po/zh-TW.po [hunk 9]
```diff
@@ -2013,127 +2022,141 @@ msgid ""
...
 
 #: src/welcome-day-1/what-is-rust.md:21
 msgid "Rust fits in the same area as C++:"
-msgstr ""
+msgstr "Rust 和 C++ 是同個領域的程式語言，具有以下特色："
```
Comment (victorhsieh) [inline po/zh-TW.po:2133]:
```suggestion
msgstr "Rust 適合用於與 C++ 同樣的領域，且具有以下特色："
```

File: po/zh-TW.po [hunk 10]
```diff
@@ -2184,13 +2207,15 @@ msgstr ""
...
-msgstr ""
+msgstr "Rust 是現代的程式語言，可完整支援萬國碼等等。"
 
 #: src/hello-world.md:33
 msgid ""
 "Rust uses macros for situations where you want to have a variable number of "
 "arguments (no function [overloading](basic-syntax/functions-interlude.md))."
 msgstr ""
+"當您想使用可變數量的引數時 (亦即無任何函式[超載](basic-syntax/functions-"
+"interlude.md))，Rust 就會採用巨集。"
```
Comment (victorhsieh) [inline po/zh-TW.po:2218]: "Rust 就會採用巨集" sounds a bit strange.  How about... 可使用 Rust 的巨集?

**Review by victorhsieh — COMMENTED** [review]
Publishing my progress so far.  I'd suggest smaller patches next time, so that reviewers can process separately.

But in any case, thanks for the translation!

File: po/zh-TW.po [hunk 11]
```diff
@@ -35,145 +36,145 @@ msgstr "翻譯"
...
+#: src/SUMMARY.md:25 src/why-rust/runtime.md:1
 msgid "Runtime Guarantees"
-msgstr ""
+msgstr "執行階段保證"
 
-#: src/SUMMARY.md:25 src/why-rust/modern.md:1
+#: src/SUMMARY.md:26 src/why-rust/modern.md:1
 msgid "Modern Features"
-msgstr ""
+msgstr "現代化功能"
```
[the remark this answers is not in the document]
Comment (henrif75) [reply po/zh-TW.po:86]: Well, my chinese skills are not there yet (or anywhere 😄 ), I'll defer to other reviewers.
@edong @hueich @kuanhungchen @mingyc @johnathan79717

File: po/zh-TW.po (hunk 10 above)
Comment (henrif75) [reply po/zh-TW.po:2218]: @edong @hueich @kuanhungchen @mingyc @johnathan79717

**Comment by henrif75** [comment]
Hi @victorhsieh ,
The check is failing probably because you're using a @gmail.com email that didn't accept the CLA, would like to check it out?
Thanks!

File: po/zh-TW.po [hunk 12]
```diff
@@ -3054,36 +3163,41 @@ msgstr ""
...
+"函式主體 (或任何區塊) 中的最後一個運算式會成為回傳值。您只要省略運算式結尾的 "
+"`;` 即可。"
 
 #: src/basic-syntax/functions.md:38
 msgid ""
 "Some functions have no return value, and return the 'unit type', `()`. The "
 "compiler will infer this if the `-> ()` return type is omitted."
 msgstr ""
+"某些函式沒有回傳值，會傳回 `()` 這個「單位型別」。如果省略 `-> ()` 傳回型別，"
+"編譯器就會推斷出這點。"
```
Comment (victorhsieh) [inline po/zh-TW.po:3188]:
```suggestion
"編譯器則會推斷出這點。"
```

**Review by victorhsieh — COMMENTED** [review]
/me making some progress slowly...

File: po/zh-TW.po [hunk 13]
```diff
@@ -597,237 +598,237 @@ msgstr "呼叫不安全的函式"
...
+#: src/SUMMARY.md:224
 msgid "PACs"
-msgstr ""
+msgstr "PAC"
 
-#: src/SUMMARY.md:218
+#: src/SUMMARY.md:225
 msgid "HAL Crates"
-msgstr ""
+msgstr "HAL Crate"
```
[the remark this answers is not in the document]
Comment (henrif75) [reply po/zh-TW.po:722]: Do you mean the plural?

File: po/zh-TW.po [hunk 14]
```diff
@@ -2297,87 +2336,102 @@ msgid ""
...
 msgstr ""
+"您也可能會不小心使用 `Rc` 或 `Arc` 建立[參照循環](https://doc.rust-lang.org/"
+"book/ch15-06-reference-cycles.html)。"
 
 #: src/why-rust/compile-time.md:25
 msgid ""
 "In fact, some will consider infinitely populating a collection a memory leak "
 "and Rust does not protect from those."
 msgstr ""
+"事實上，有些人會認為無限地填充集合將導致記憶體流失，而 Rust 並不能避免這種情"
```
[the remark this answers is not in the document]
Comment (henrif75) [reply po/zh-TW.po:2417]: Can you make a suggestion?

File: po/zh-TW.po [hunk 15]
```diff
@@ -2289,6 +2325,9 @@ msgid ""
 "have the memory unsafety issues. In addition, you get a modern language with "
 "constructs like pattern matching and built-in dependency management."
 msgstr ""
+"具備 C 或 C++ 經驗：Rust 會透過借用檢查器，徹底刪除一整類的__「執行階段錯"
```
[the remark this answers is not in the document]
Comment (henrif75) [reply po/zh-TW.po:2328]: I don't think it is, I'll do a batch change to remove them.

File: po/zh-TW.po [hunk 16]
```diff
@@ -3376,7 +3520,7 @@ msgstr ""
 msgid ""
 "You can print such an array by asking for its debug representation with `{:?}"
 "`:"
-msgstr ""
+msgstr "您可以使用 `{:?}` 要求陣列的偵錯表示法，藉此輸出這樣的陣列："
```
Comment (victorhsieh) [inline po/zh-TW.po:3523]:
```suggestion
msgstr "您可以使用 `{:?}` 偵錯表示法輸出這樣的陣列："
```

File: po/zh-TW.po [hunk 17]
```diff
@@ -3420,6 +3564,8 @@ msgid ""
 "and a function `transpose` which will transpose a matrix (turn rows into "
 "columns):"
 msgstr ""
+"使用上述程式碼編寫可用用於美化矩陣排版的 `pretty_print` 函式，以及會轉置矩陣 "
```
Comment (victorhsieh) [inline po/zh-TW.po:3567]:
```suggestion
"使用上述程式碼編寫可用用於美化矩陣排版的 `pretty_print` 函式，以及用於轉置矩陣 "
```

File: po/zh-TW.po [hunk 18]
```diff
@@ -3671,27 +3833,27 @@ msgid ""
...
-msgstr ""
+msgstr "另一方面，`static` 則更類似於 C++ 中的 `const` 或可變動的全域變數。"
 
-#: src/basic-syntax/static-and-const.md:50
+#: src/basic-syntax/static-and-const.md:60
 msgid ""
 "It isn't super common that one would need a runtime evaluated constant, but "
 "it is helpful and safer than using a static."
-msgstr ""
+msgstr "需要執行階段評估常數的情況雖不常見，但這會比使用靜態項目更有用且安全。"
```
Comment (victorhsieh) [inline po/zh-TW.po:3850]:
```suggestion
msgstr "需要在執行階段評估常數的情況雖不常見，但這會比使用靜態項目更有用且安全。"
```

File: po/zh-TW.po [hunk 19]
```diff
@@ -5530,15 +5696,16 @@ msgstr ""
 msgid "Structs, enums, methods."
 msgstr ""
 
-#: src/welcome-day-2.md:7
```
Comment (victorhsieh) [inline po/zh-TW.po:5533]: Since this and a few below change it to day 1, I assume these are fixes.

File: po/zh-TW.po [hunk 20]
```diff
@@ -6689,23 +6864,27 @@ msgid ""
...
 "errors when running the code. Point out the places where your code is now "
 "inexhaustive and how the compiler tries to give you hints."
 msgstr ""
+"您可以嘗試在列舉定義中加入第三個變體，並在執行程式碼時顯示錯誤。請向學員指出"
+"程式碼現在有哪些地方還不詳盡，並說明編譯器會如何嘗試給予提示。"
 
 #: src/pattern-matching/destructuring-structs.md:3
 msgid "You can also destructure `structs`:"
-msgstr ""
+msgstr "您也可以解構 `structs`："
```
Comment (victorhsieh) [inline po/zh-TW.po:6887]:
```suggestion
msgstr "您也可以解構 `struct`："
```

File: po/zh-TW.po [hunk 21]
```diff
@@ -21350,3 +21540,518 @@ msgid ""
...
+"rust-lang.org/cargo/reference/registries.html), git, folders, and more."
+msgstr ""
+"您也可以透過其他[註冊資料庫](https://doc.rust-lang.org/cargo/reference/"
+"registries.html)、git、資料夾等管道解析依附元件。"
+
+#: src/cargo/running-locally.md:15
+msgid ""
+"You can use any later version too since Rust maintains backwards "
+"compatibility."
+msgstr "由於 Rust 保有回溯相容性，因此您也可以使用任何後續版本。"
```
Comment (victorhsieh) [inline po/zh-TW.po:21681]:
```suggestion
msgstr "由於 Rust 保有回溯相容性，您也可以使用任何後續版本。"
```

File: po/zh-TW.po [hunk 22]
```diff
@@ -21350,3 +21540,518 @@ msgid ""
...
+#: src/hello-world/small-example.md:44
+msgid ""
+"In a shell `rustup doc std::fmt` will open a browser on the local std::fmt "
+"documentation"
+msgstr ""
+"在殼層中，`rustup doc std::fmt` 會開啟本機 std::fmt 說明文件上的瀏覽器。"
+
+#: src/why-rust/runtime.md:6
+msgid "Integer overflow is defined (panic or wrap-around)."
+msgstr "定義了整數溢位現象 (恐慌或迴繞)。"
```
Comment (victorhsieh) [inline po/zh-TW.po:21731]:
```suggestion
msgstr "整數溢位的行為是明確的 (恐慌或迴繞)。"
```

**Review by victorhsieh — APPROVED** [review]
LGTM with nits.  Thanks for the translation!

**Pull request merged**